### PR TITLE
Added I18N utility

### LINKED ISSUE #3: Add I18NUtility
- [x] <b>Description</b> <br>
    Add utility to internationalize strings.
- [x] <b>Use cases</b> <br>
    1. Ability to add/remove I18N file to search.
    2. Search and retrieve a specific I18N string across all specified I18N files.
- [x] <b>Proposed design</b> <br>
    Migrate I18NUtility from https://github.com/padaiyal/jIcing.
- [x] <b>Alternative design approaches</b> <br>
   N/A
- [x] <b>Workaround</b> <br>
    N/A
- [x] <b>Additional information</b> <br>
    N/A

### PR DESCRIPTION
**Pull request by silvs110**
## Description
The added utility class allows to add/remove properties files used for internationalization.
It also retrieves values from all specified I18N files.

Closes #3.

## Author's Checklist
- [x] **These changes don't break existing dependants.** If they do, and you suggest to proceed, 
please explain why.

## Reviewer's Checklist
- [x] **All classes have documentation comments.**
- [x] **All methods have documentation comments.**
- [x] **No parameters are hardcoded.** Should be loaded as properties instead. If there needs to be hardcoded properties, explain why.
- [x] **All information to be logged/displayed to the user are internationalized.** If not, explain why. 
- [x] **README.md has been updated if needed.**

## Assignee's Checklist
- [x] **All GitHub action checks have passed.**
- [x] **All reviewers have approved.**
- [x] **Relevant documentation has been updated if needed.**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -14,7 +14,15 @@
     <version>2021.01.09</version>
```
Comment (the-c0d3br34k3r) [inline pom.xml:14]: Change artifact version.

**Review by the-c0d3br34k3r — CHANGES_REQUESTED** [review]
Changes requested.

File: pom.xml (hunk 1 above)
Comment (silvs110) [reply pom.xml:14]: currently jPopper submodule uses the 2021.01.09 version

Comment (the-c0d3br34k3r) [reply pom.xml:14]: Gotcha!

File: src/main/java/org/padaiyal/utilities/I18nUtility.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Utility class for internationalizing strings.
+ */
+public final class I18nUtility {
+
+  /**
+   * logger for I18nUtility class.
```
Comment (the-c0d3br34k3r) [inline src/main/java/org/padaiyal/utilities/I18nUtility.java]:
```suggestion
   * Logger for I18nUtility class.
```

File: src/test/java/org/padaiyal/utilities/I18nUtilityTest.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+   * @param resourceBundleName The resource bundle to retrieve the key from.
+   * @param key                The key to test.
+   * @param message            The expected value corresponding to the provided key.
+   * @param exceptionTypeName  The expected exception type. This is null if the key is valid.
+   */
+  @ParameterizedTest
+  @CsvSource({
+      "I18N_Resource_Bundle, com.sample.message, Hello All!!!, ",
+      "I18N_Resource_Bundle_invalid, com.sample.message, , MissingResourceException.class",
+      // Invalid resource bundle.
```
Comment (the-c0d3br34k3r) [inline src/test/java/org/padaiyal/utilities/I18nUtilityTest.java]: Put comment before input.

File: src/test/java/org/padaiyal/utilities/I18nUtilityTest.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+   * java.lang.Object...) with valid and invalid keys and values.
+   *
+   * @param key               The key to test.
+   * @param value             The value to format the message.
+   * @param formattedMessage  The expected formatted message.
+   * @param exceptionTypeName The expected exception type. This is null if the key is valid.
+   */
+  @ParameterizedTest
+  @CsvSource({
+      "com.sample.formattedmessage, value1, Hello value1!!!, ", // Valid key and valid value.
```
Comment (the-c0d3br34k3r) [inline src/test/java/org/padaiyal/utilities/I18nUtilityTest.java]: Put comment before input.

File: src/test/java/org/padaiyal/utilities/I18nUtilityTest.java [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    } else {
+      Assertions.assertEquals(
+          formattedMessage,
+          I18nUtility.getFormattedStringFromResourceBundle(key, resourceBundleName, value)
+      );
+    }
+  }
+
+  /**
+   * Tests org.padaiyal.utilities.I18nUtility ::addResourceBundle(java.lang.Class, java.lang.String,
```
Comment (the-c0d3br34k3r) [inline src/test/java/org/padaiyal/utilities/I18nUtilityTest.java]:
```suggestion
   * Tests org.padaiyal.utilities.I18nUtility::addResourceBundle(java.lang.Class, java.lang.String,
```

**Review by the-c0d3br34k3r — CHANGES_REQUESTED** [review]
Requested changes.

File: src/main/java/org/padaiyal/utilities/I18nUtility.java (hunk 2 above)
Comment (silvs110) [reply src/main/java/org/padaiyal/utilities/I18nUtility.java]: done.

File: src/test/java/org/padaiyal/utilities/I18nUtilityTest.java (hunk 4 above)
Comment (silvs110) [reply src/test/java/org/padaiyal/utilities/I18nUtilityTest.java]: done.

File: src/main/java/org/padaiyal/utilities/I18nUtility.java [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+      throw new MissingResourceException(
+          String.format("Can't find resource bundle %s", resourceBundleName),
+          I18nUtility.class.getName(),
+          resourceBundleName
+      );
+    }
+  }
+
+  /**
+   * Gets a the added resource bundle names.
```
Comment (the-c0d3br34k3r) [inline src/main/java/org/padaiyal/utilities/I18nUtility.java]:
```suggestion
   * Gets a list of added resource bundle names.
```

File: src/test/java/org/padaiyal/utilities/I18nUtilityTest.java [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+        .filter(resourceBundleName ->
+            !resourceBundleName.equals(defaultResourceBundle)
+        )
+        .forEach(I18nUtility::removeResourceBundle);
+  }
+
+  /**
+   * Test the org.padaiyal.utilities.I18nUtility::getString(java.lang.String) with valid keys.
+   *
+   * @param key The key to test.
```
Comment (the-c0d3br34k3r) [inline src/test/java/org/padaiyal/utilities/I18nUtilityTest.java:62]: Missing message parameter.

File: src/test/java/org/padaiyal/utilities/I18nUtilityTest.java [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+    locale = new Locale(language, region);
+    defaultResourceBundle = PropertyUtility.getProperty("I18NUtilityTest.resourcebundle.default");
+    I18nUtility.addResourceBundle(I18nUtilityTest.class, defaultResourceBundle, locale);
+  }
+
+  /**
+   * Removes all added resource bundle except for the default resource bundle.
+   */
+  @AfterEach
+  void testTearDown() {
```
Comment (the-c0d3br34k3r) [inline src/test/java/org/padaiyal/utilities/I18nUtilityTest.java]:
```suggestion
  void tearDown() {
```
The `test` prefix should only be used for actual tests. Not setting up or tearing down dependencies.

Comment (the-c0d3br34k3r) [reply src/test/java/org/padaiyal/utilities/I18nUtilityTest.java]: Why is the setUp() set to @BeforeAll, but the teardown is set to @AfterEach ?

File: src/test/java/org/padaiyal/utilities/I18nUtilityTest.java [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+        );
+    String actualMessage = I18nUtility.getString(key);
+
+    Assertions.assertEquals(message, actualMessage);
+  }
+
+  /**
+   * Test the org.padaiyal.utilities.I18nUtility::getString(java.lang.String) with invalid keys.
+   *
+   * @param key The key to test.
```
Comment (the-c0d3br34k3r) [inline src/test/java/org/padaiyal/utilities/I18nUtilityTest.java:87]: Missing exceptionTypeClass parameter.

File: src/test/java/org/padaiyal/utilities/I18nUtilityTest.java [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+  @ParameterizedTest
+  @CsvSource({
+      // Valid key and valid value.
+      "I18N_Resource_Bundle, com.sample.formattedmessage, value1, Hello value1!!!",
+      // Valid key with null value type.
+      "I18N_Resource_Bundle, com.sample.formattedmessage1, , Hello null!!!",
+  })
+  void testGetFormattedStringFromOneResourceBundleWithValidInputs(String resourceBundleName,
+      String key,
+      String value, String formattedMessage) {
```
Comment (the-c0d3br34k3r) [inline src/test/java/org/padaiyal/utilities/I18nUtilityTest.java]: Inconsistent formatting.

**Review by the-c0d3br34k3r — CHANGES_REQUESTED** [review]
Requested changes.

File: src/test/java/org/padaiyal/utilities/I18nUtilityTest.java (hunk 8 above)
Comment (silvs110) [reply src/test/java/org/padaiyal/utilities/I18nUtilityTest.java]: @BeforeAll is used to setup before the tests in I18nUtilityTests, but @AfterEach just cleans up after each test has ran.

File: src/test/java/org/padaiyal/utilities/I18nUtilityTest.java (hunk 9 above)
Comment (silvs110) [reply src/test/java/org/padaiyal/utilities/I18nUtilityTest.java:87]: done.

File: src/test/java/org/padaiyal/utilities/I18nUtilityTest.java (hunk 7 above)
Comment (silvs110) [reply src/test/java/org/padaiyal/utilities/I18nUtilityTest.java:62]: done.

File: src/main/java/org/padaiyal/utilities/I18nUtility.java (hunk 6 above)
Comment (silvs110) [reply src/main/java/org/padaiyal/utilities/I18nUtility.java]: done.

File: src/test/java/org/padaiyal/utilities/I18nUtilityTest.java (hunk 10 above)
Comment (silvs110) [reply src/test/java/org/padaiyal/utilities/I18nUtilityTest.java]: done.

**Review by the-c0d3br34k3r — APPROVED** [review]

**Pull request merged**